std/json.c3

```diff
@@ -0,0 +1,428 @@
+    }
+
+    const next_token : u8 = optional_next_token:data();
+
+    if next_token == char("\"") {
+        return cursor:parse_string();
+    }
+
+    if next_token == char("{") {
+        return cursor:parse_object();
+    }
+
+    if next_token == char("[") {
+        return cursor:parse_array();
+    }
+
+    // Numbers don't have a prefix character, so just try and parse them.
+    let node : Optional<JSON_Node> = cursor:parse_number();
+    if node:has_value() {
+        return node;
+    }
+
+    if cursor:eat(sv("null")) {
+        return make_optional(new_json_node(/* null */ 0));
+    }
+
+    if cursor:eat(sv("true")) {
+        return make_optional(new_json_node(/* true */ 1));
+    }
+
+    if cursor:eat(sv("false")) {
+        return make_optional(new_json_node(/* false */ 2));
+    }
+
+    return make_optional<JSON_Node>();
+}
+
+function parse_element : (cursor : Cursor&) -> Optional<JSON_Node> = {
+    cursor:eat_whitespace();
+
+    const value_node : Optional<JSON_Node> = cursor:parse_value();
+
+    cursor:eat_whitespace();
+
+    return value_node;
+}
+
+function parse_string : (cursor : Cursor&) -> Optional<JSON_Node> = {
+    if !cursor:eat(char("\"")) {
+        return make_optional<JSON_Node>();
+    }
+
+    const start_pos : isize = cursor.position;
+
+    while !cursor:is_eof() {
+        if cursor:eat(char("\"")) {
+            const end_pos : isize = cursor.position - 1;
+
+            let node : JSON_Node = new_json_node(/* string */ 3);
+            node.value = make_optional(
+                cursor.data:slice_between(start_pos, end_pos)
+            );
+
+            return make_optional(node);
+        }
+
+        // TODO escape characters.
+
+        cursor:next();
+    }
+
+    return make_optional<JSON_Node>();
+}
+
+function is_digit_helper : (token : Optional<u8>) -> bool = {
+    return token:has_value() and ascii_is_digit(token:data());
+}
+
+function parse_number : (cursor : Cursor&) -> Optional<JSON_Node> = {
+    const start_pos : isize = cursor.position;
+    let is_valid : bool = false;
+
+    // This is optional, so discard the return value.
+    cursor:eat(char("-"));
+
+    if !cursor:eat(char("0")) {
+        while is_digit_helper(cursor:peak()) {
+            is_valid = true;
+            cursor:next();
+        }
+    }
+    else {
+        is_valid = true;
+    }
+
+    // TODO parse fraction.
+    // TODO parse exponent.
+
+    const end_pos : isize = cursor.position;
+
+    if !is_valid {
+        return make_optional<JSON_Node>();
+    }
+
+    let node : JSON_Node = new_json_node(/* string */ 4);
+    node.value = make_optional(cursor.data:slice_between(start_pos, end_pos));
+
+    return make_optional(node);
+}
+
+// TODO this and write below should be the only public functions.
+function json_parse : (source : StringView) -> Optional<JSON_Node> = {
+    // FIXME memory safety, memory leaks.
+    let cursor : Cursor = {
+        source, 0, initialize_allocator()
+    };
```

Comment:
Lol, leaking the entire allocator is kinda funny tbh. Thats actually the worst possible leak :laughing:

Reply:
Yeah, our memory management is pretty bad :(

Reply:
I really want a global allocator though